File: 234-palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        ls=[]
        while head.next:
            ls.append(head.val)
            head=head.next
        ls.append(head.val)    
        return ls==ls[::-1]

File: 234-palindrome-linked-list/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* dummy = NULL;
        while(head){
            ListNode* next = head->next;
            head->next = dummy;
            dummy = head;
            head = next;
        }
        return dummy;
    }
    bool isPalindrome(ListNode* head) {
        ListNode* slow = head, *fast = head;
        while(fast->next and fast->next->next){
            slow=slow->next;
            fast=fast->next->next;
        }
        // cout<<slow->val<<endl;
        ListNode*newhead = reverseList(slow);
        // cout<<newhead->val<<endl;
        fast = head;
        while(fast and newhead){

            if(fast->val != newhead->val){
                return false;
            }
            fast=fast->next;
            newhead = newhead->next;
        }
        return true;
    }
};